Qt/PropertiesDialog: Add "Game Config" tab

spycrab opened the pull request:
![Screenshot](https://i.imgur.com/0qzs5wb.png)

spycrab commented:
@MayImilae @JMC47

BhaaLseN commented:
Any thoughts on making those determinism string values `constexpr` as well?

spycrab commented:
Sounds reasonable.

lioncash commented on Source/Core/DolphinQt2/Config/GameConfigWidget.h:
```diff
@@ -0,0 +1,74 @@
+  void EditUserConfig();
+  void ViewDefaultConfig();
+
+  void LoadCheckBox(QCheckBox* checkbox, const std::string&& section, const std::string&& key);
```
Do const rvalues really provide a benefit here over const references?

lioncash commented on Source/Core/DolphinQt2/Config/GameConfigWidget.h:
```diff
@@ -0,0 +1,74 @@
+#include <QWidget>
+
+#include "Common/IniFile.h"
+#include "DolphinQt2/GameList/GameFile.h"
```
As far as I'm aware, GameFile can be forward declared.

lioncash commented on Source/Core/DolphinQt2/Config/GameConfigWidget.h:
```diff
@@ -0,0 +1,74 @@
+
+#include <string>
+
+#include <QSpinBox>
```
QSpinBox is already forward declared further down.

spycrab commented:
@lioncash All fixed now.

MayImilae commented:
Ok, took a look at this window! I'll start with macOS.

![screen shot 2018-03-04 at 7 18 16 pm](https://user-images.githubusercontent.com/6551020/36956077-26e4a208-1fe1-11e8-93ee-73eaf47aba69.png)

This is something of a common problem with Qt right now in macOS - everything is **HUGE!!** The spacing is all wrong and the window is too big for my poor laptop's screen. Considering none of these spacing problems exist on the linux screenshots, I'm assuming this is just a macOS issue that needs to be ironed out. Desperately, it's bad.

...or maybe it's on windows too, ok.

![spacing windows](https://user-images.githubusercontent.com/6551020/36956228-0db439fa-1fe2-11e8-8efe-7736fc07070c.png)

The spacing is all wrong here too. Not as bad, I mean, it isn't suuuuper wide as well as super tall, but it's pretty overly spaced. Note that unlike macOS, I can make it smaller, but by default it is at the size shown above. Even when shrunken to the minimum though, there's still wasted space. Around the text at the top and such, see below. Note that if I close the window and reopen it it is at the bigger size above.

![spacing windowsshrunken](https://user-images.githubusercontent.com/6551020/36956713-b279a31a-1fe4-11e8-8070-47a33b986ff1.png)

This is not related to this PR so I won't go into detail here, but there are spacing problems elsewhere too. Spacing issues are kind of a big problem with Qt right now, with tons of things in Windows and macOS (especially) just having waisted space for no apparent reason. Yay platform specific issues! We'll need to get that ironed out before we make the switch. I'll talk with you over IRC when you're available, I've written up a list! :3

Since the spacing problems are so widespread, if you want to just merge this and fix all the spacing issues together, I'd be fine with that. 

Anyway, aside from the spacing issues... I'm not what the wii widescreen option is doing here? I mean, sure it would have been useful years ago, but all of that was handled with the heuristic and gameinis now. Wii games that need to be 4:3 only, like virtual console titles, have that already set in the gameinis dolphin ships with. So there isn't a reason to promote that for easier user access. Why did you add it?

Also the buttons at the bottom of the game config tab (edit user config and view default config) are too tight could use more space around the text. I would see if like that's on macOS too, but I can't see them!

spycrab commented:
> Anyway, aside from the spacing issues... I'm not what the wii widescreen option is doing here? I mean, sure it would have been useful years ago, but all of that was handled with the heuristic and gameinis now. Wii games that need to be 4:3 only, like virtual console titles, have that already set in the gameinis dolphin ships with. So there isn't a reason to promote that for easier user access. Why did you add it?

Removed.

> Also the buttons at the bottom of the game config tab (edit user config and view default config) are too tight could use more space around the text. I would see if like that's on macOS too, but I can't see them!

Fixed.

I will address the other "general" spacing issues in another PR.

spycrab commented:
@dolphin-emu-bot rebuild

leoetlino commented on Source/Core/DolphinQt2/Config/GameConfigWidget.cpp:
```diff
@@ -0,0 +1,412 @@
+
+  if (m_gameini_local.GetOrCreateSection(section)->Get(key, &checked))
+    return checkbox->setCheckState(checked ? Qt::Checked : Qt::Unchecked);
+  else if (m_gameini_default.GetOrCreateSection(section)->Get(key, &checked))
```
You don't need an `else` after a `return`.

leoetlino commented on Source/Core/DolphinQt2/Config/GameConfigWidget.cpp:
```diff
@@ -0,0 +1,412 @@
+
+  m_gameini_local.GetOrCreateSection(section)->Set(key, checked);
+
+  return;
```
This return doesn't seem to serve any purpose, could you remove it?

leoetlino commented on Source/Core/DolphinQt2/Config/GameConfigWidget.cpp:
```diff
@@ -0,0 +1,412 @@
+    if (!m_gameini_default.GetIfExists("Core", "GPUDeterminismMode", &determinism_mode))
+      determinism_mode = DETERMINISM_NOT_SET_STRING;
+
+  if (determinism_mode == DETERMINISM_AUTO_STRING)
```
IMO, a switch statement would be better here.

spycrab replied:
A switch on a string?

leoetlino replied:
Oh whoops, disregard what I said.

leoetlino commented on Source/Core/DolphinQt2/Config/GameConfigWidget.cpp:
```diff
@@ -0,0 +1,412 @@
+  }
+
+  // Stereoscopy
+  int depth_percentage = 100;
```
There is quite a lot of code duplication here (there are 3 calls to `setValue` for one variable). I'd just do:

```
int depth_percentage = 100;
m_gameini_default.GetIfExists(...);
m_gameini_local.GetIfExists(...);
m_depth_slider->setValue(depth_percentage);
```
